src/fabxdevice.h

```diff
@@ -296,12 +318,130 @@ inline void FabXDevice::loop() {
           break;
         }
       }
+
+      // Tool not found, e.g. backend configuration was changed
+      if (mCurrentState == States::TOOL_SELECT) {
+        X_DEBUG("Tool %s not found", mAuthorizedToolIds.ToolIds[mSelectedTool].c_str());
+        mCurrentState = States::IDLE;
+      }
     }
     break;
-  case States::REQUEST_SECOND_FACTOR:
-    X_DEBUG("Second Factor");
-    mCurrentState = States::IDLE;
+  case States::REQUEST_SECOND_FACTOR: {
+    X_DEBUG("Requesting 2FA from keypad");
+    mKeypad->setCommand(Keypad::Command::CODE_REQUIRED);
+    int stateWasEnterCodeCount = 0;
+    Keypad::State state = mKeypad->getState();
+    Keypad::State prevState = state;
+    bool internalTimeout = true;
+    unsigned long now = millis();
+    while (millis() < now + 10000)
+    {
+      state = mKeypad->getState();
+      if (prevState != state && state == Keypad::State::TYPING) {
+        X_DEBUG("User is typing");
```

Comment:
We could make the timeout aware of user typing, as it is in the keypad firmware.